File: trunk/apps/oslc-km-srl/src/main/java/com/reusecompany/oslc/km/srl/Artifact.java
package com.reusecompany.oslc.km.srl;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import org.eclipse.lyo.oslc4j.core.annotation.OslcDescription;
import org.eclipse.lyo.oslc4j.core.annotation.OslcName;
import org.eclipse.lyo.oslc4j.core.annotation.OslcNamespace;
import org.eclipse.lyo.oslc4j.core.annotation.OslcOccurs;
import org.eclipse.lyo.oslc4j.core.annotation.OslcPropertyDefinition;
import org.eclipse.lyo.oslc4j.core.annotation.OslcRange;
import org.eclipse.lyo.oslc4j.core.annotation.OslcReadOnly;
import org.eclipse.lyo.oslc4j.core.annotation.OslcRepresentation;
import org.eclipse.lyo.oslc4j.core.annotation.OslcResourceShape;
import org.eclipse.lyo.oslc4j.core.annotation.OslcTitle;
import org.eclipse.lyo.oslc4j.core.annotation.OslcValueShape;
import org.eclipse.lyo.oslc4j.core.annotation.OslcValueType;
import org.eclipse.lyo.oslc4j.core.model.AbstractResource;
import org.eclipse.lyo.oslc4j.core.model.Occurs;
import org.eclipse.lyo.oslc4j.core.model.OslcConstants;
import org.eclipse.lyo.oslc4j.core.model.Representation;
import org.eclipse.lyo.oslc4j.core.model.ValueType;

import com.hp.hpl.jena.rdf.model.SimpleSelector;
import com.hp.hpl.jena.rdf.model.StmtIterator;
import com.reusecompany.oslc.km.vocabs.SRL;

@OslcNamespace(SRL.NAMESPACE)
@OslcName(SRL.ARTIFACT)
@OslcResourceShape(title = "SRL Artifact Resource Shape", 
describes = SRL.NAMESPACE + SRL.ARTIFACT )
public class Artifact extends AbstractResource {
	public static final Artifact EMPTY_ARTIFACT = new Artifact("empty");
	/**Properties as KE**/
	Term term;
	boolean ke = false;
	
	/**Properties as Artifact**/
	ArtifactType artifactType = DEFAULT_ARTIFACT_TYPE;
	List<RSHP> rshps = new LinkedList<RSHP>();
	List<MetaProperty> metaproperties = new LinkedList<MetaProperty>();
	List<Artifact> ownedArtifacts = new LinkedList<Artifact>();
	
	/**Properties to access an Artifact**/
	URI fromURI = null;
	URI sparqlEndpoint = null;
	String rdfContent = null;
	
	/**OSLC Core properties**/
	private URI serviceProvider;
	private String identifier = "";
	private String title = "";
	private String description = "";
	private URI preferredVisualization;
	
	public static ArtifactType DEFAULT_ARTIFACT_TYPE = null;
	
	static{
		DEFAULT_ARTIFACT_TYPE = new ArtifactType();
		try {
			DEFAULT_ARTIFACT_TYPE.setAbout(new URI(SRL.NAMESPACE));
		} catch (URISyntaxException e) {
		
		}
		DEFAULT_ARTIFACT_TYPE.setIdentifier("19");
		DEFAULT_ARTIFACT_TYPE.setName("DEFAULT");
	}
	
	public Artifact() {
	}
	
	public Artifact(Term term) {
		this.term = term;
	}

	public Artifact(String term) {
		this.term = new Term();
		this.term.setPrefLabel(term);
	}


	public Artifact(boolean isKE, URI uri) {
		this.ke = isKE;
		this.setAbout(uri);
	}






	@OslcDescription("The scope of a resource is a URI for the resource's OSLC Service Provider.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(OslcConstants.OSLC_CORE_NAMESPACE + "serviceProvider")
	@OslcRange(OslcConstants.TYPE_SERVICE_PROVIDER)
	@OslcReadOnly
	@OslcTitle("Service Provider")
	public URI getServiceProvider() {
		return serviceProvider;
	}
	public void setServiceProvider(URI value) {
		this.serviceProvider = value;
	}


	@OslcDescription("A unique identifier for a resource. Assigned by the service provider when a resource is created. Not intended for end-user display.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(OslcConstants.DCTERMS_NAMESPACE + "identifier")
	@OslcReadOnly
	@OslcTitle("Identifier")
	public String getIdentifier() {
		return identifier;
	}

	public void setIdentifier(String identifier) {
		this.identifier = identifier;
	}



	@OslcDescription("The resource type URIs.")
	@OslcName("type")
	@OslcPropertyDefinition(OslcConstants.RDF_NAMESPACE + "type")
	@OslcTitle("Types")
	public URI[] getRdfTypes() {
		return this.getTypes().toArray(new URI[this.getTypes().size()]);
	}

	public void setRdfTypes(URI[] rdfTypes) {
		this.getTypes().addAll(Arrays.asList(rdfTypes));
	}


	@OslcDescription("The resource type URIs.")
	@OslcOccurs(Occurs.ZeroOrMany)
	@OslcName("owned-artifacts")
	@OslcPropertyDefinition(SRL.NAMESPACE + "owned-artifacts")
	@OslcTitle("List of subartifacts")
	public Artifact[] getOwnedArtifacts() {
		return this.ownedArtifacts.toArray(new Artifact[this.ownedArtifacts.size()]);
	}

	public List<Artifact> getOwnedArtifactsAsList() {
		return this.ownedArtifacts;
	}
	
	public void setOwnedArtifactsAsList( List<Artifact> ownedArtifacts) {
		if(ownedArtifacts != null){
			this.ownedArtifacts = ownedArtifacts;
		}
	}
	
	public void setOwnedArtifacts(Artifact[] ownedArtifacts) {
		this.ownedArtifacts.addAll(Arrays.asList(ownedArtifacts));
	}
	
	@OslcDescription("The visualization of this artifact.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "preferred-visualization")
	@OslcReadOnly
	@OslcName("preferred-visualization")
	@OslcTitle("Graphical view of this artifact.")
	public URI getPreferredVisualization() {
		//FIXME:
//		URI outURI;
//		if (ImageLoader.pictures.TrygetValue(this.identifier, out outURI)) {
//			return outURI;
//		}
		return this.preferredVisualization;
	}

	public void setPreferredVisualization(URI URI) {
		this.preferredVisualization = URI;
	}


	@OslcDescription("The URI from which external content is read.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "from-URI")
	@OslcReadOnly
	@OslcName("from-URI")
	@OslcTitle("URI to read external content.")
	public URI getFromURI() {
		return this.fromURI;
	}

	public void setFromURI(URI URI){
		this.fromURI = URI;
	}



	@OslcDescription("The SPARQL endpoint from which external content is read.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "sparql-endpoint")
	@OslcReadOnly
	@OslcTitle("URI to read external content.")
	public URI getSparqlEndpoint() {
		return this.sparqlEndpoint;
	}

	public void setSparqlEndpoint(URI URI) {
		this.sparqlEndpoint = URI;
	}


	@OslcDescription("The RDF content of a resource.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "value")
	@OslcReadOnly
	@OslcTitle("The RDF content of a resource.")
	public String getRdfContent() {
		return this.rdfContent;
	}

	public void setRdfContent(String rdfContent) {
		this.rdfContent = rdfContent;
	}

	@OslcDescription("The title of an artifact.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "title")
	@OslcReadOnly
	@OslcTitle("The title of an artifact.")
	public String getTitle() {
		return this.title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	@OslcDescription("The description of an artifact.")
	@OslcOccurs(Occurs.ExactlyOne)
	@OslcPropertyDefinition(SRL.NAMESPACE + "description")
	@OslcReadOnly
	@OslcTitle("The description of an artifact.")
	public String getDescription() {
		return this.description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	@OslcDescription("Relationships")
	@OslcPropertyDefinition(SRL.NAMESPACE + "rshps")
	@OslcReadOnly
	@OslcRepresentation(Representation.Inline)
	@OslcTitle("Relationships")
	@OslcName("rshps")
	@OslcValueType(ValueType.LocalResource)
	public RSHP[] getRshps() {
		return this.rshps.toArray(new RSHP[this.rshps.size()]);
	}
	public void setRshps(RSHP[] rshps) {
		this.rshps.clear();
		this.rshps.addAll(Arrays.asList(rshps));
	}

	public void setRshpsAsList(List<RSHP> rshps) {
		this.rshps.clear();
		this.rshps = null;
		this.rshps = rshps;
	}


	public List<RSHP> getRshpsAsList() {
		return this.rshps;
	}

	@OslcDescription("Artifact Type")
	@OslcPropertyDefinition(SRL.NAMESPACE + "artifact-type")
	@OslcReadOnly
	@OslcRepresentation(Representation.Inline)
	@OslcTitle("Artifact Type")
	@OslcName("artifact-type")
	@OslcValueType(ValueType.LocalResource)
	public ArtifactType getArtifactType() {
		return this.artifactType;
	}

	public void setArtifactType(ArtifactType artifactType) {
		this.artifactType = artifactType;
	}


	@OslcDescription("Artifact term")
	@OslcPropertyDefinition(SRL.NAMESPACE + "term")
	@OslcReadOnly
	@OslcRepresentation(Representation.Inline)
	@OslcTitle("Artifact term")
	@OslcValueType(ValueType.LocalResource)
	public Term getTerm() {
		return term;
	}

	public void setTerm(Term term) {
		this.term = term;
	}

	@OslcDescription("The list of metaproperties in this knowledge element.")
	@OslcPropertyDefinition(SRL.NAMESPACE + "metaproperties")
	@OslcReadOnly
	@OslcRepresentation(Representation.Inline)
	@OslcTitle("The Metaproperties")
	@OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)
	@OslcValueType(ValueType.LocalResource)
	public MetaProperty[] getMetaproperties() {
		return this.metaproperties.toArray(new MetaProperty[this.metaproperties.size()]);
	}
	public void setMetaproperties(MetaProperty[] metaproperties) {
		this.metaproperties.clear();
		this.metaproperties.addAll(Arrays.asList(metaproperties));
	}

	public List<MetaProperty> getMetaPropertiesAsList() {
		return this.metaproperties;
	}
	public void setMetapropertiesAsList(List<MetaProperty> metaproperties) {
		this.metaproperties = metaproperties;
	}

	public boolean isKE() {
		return this.ke;
	}

	@Override
	public String toString() {
		return "Artifact, uri: "+(this.getAbout()!=null?this.getAbout().toString():"")+", [term=" + term + "], "
				+ "artifactType=" + artifactType + ", rshps=" + rshps + ", metaproperties="
				+ metaproperties+ "]";
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((term == null) ? 0 : term.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Artifact other = (Artifact) obj;
		if (term == null) {
			if (other.term != null)
				return false;
		} else if (!term.equals(other.term))
			return false;
		return true;
	}


	


	
}

File: trunk/apps/dotnet/OSLC-KM-SRL/srl/Artifact.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using Newtonsoft.Json;
using OSLC4Net.Core.Attribute;
using OSLC4Net.Core.Model;

namespace OSLC_KM_SRL.srl {
    [OslcNamespace(SRLShapeConstants.OSLC_KM_SRL_DOMAIN)]
    [OslcName(SRLShapeConstants.SRL_ARTIFACT)]
    [OslcResourceShape(title = "SRL Artifact Resource Shape", describes = new string[] {
        SRLShapeConstants.SRL_ARTIFACT_URI})]
    public class Artifact : AbstractResource {

        #region Properties for KE
        [JsonProperty(PropertyName = "term")]
        private Term term;
        [JsonProperty(PropertyName = "ke")]
        private bool ke = false;

        #endregion

        #region Properties for Artifact

        [JsonProperty(PropertyName = "artifactType")]
        private Type type;
        [JsonProperty(PropertyName = "rshps")]
        private List<Relationship> relationships = null;
        [JsonProperty(PropertyName = "metadatas")]
        private List<MetaData> metadata = null;
        [JsonProperty(PropertyName = "data")]
        private List<Data> data = null;
        [JsonProperty(PropertyName = "ownedArtifacts")]
        private List<Artifact> ownedArtifacts = null;


        #endregion

        #region OSLC Core properties
        [JsonProperty(PropertyName = "identifier")]
        private string identifier = string.Empty;
        [JsonProperty(PropertyName = "serviceProvider")]
        private Uri serviceProvider;
        #endregion

        #region Image, Binary and Physical Path
        [JsonProperty(PropertyName = "Content")]
        private string Content = null;
        [JsonProperty(PropertyName = "Snapshot")]
        private string SnapShot = null;
        [JsonProperty(PropertyName = "PhysicalPath")]
        private string PhysicalPath = string.Empty;
        #endregion

        #region Constructors
        public Artifact() {
            init();
        }
        public Artifact(bool isKE, Uri uri) {
            this.ke = isKE;
            this.SetAbout(uri);
            init();
        }
        public Artifact(string term) {
            this.term = new Term();
            this.term.SetPrefLabel(term);
            init();
        }
        private void init() {
            this.type = DEFAULT_ARTIFACT_TYPE;
            this.relationships = new List<Relationship>();
            this.metadata = new List<MetaData>();
            this.data = new List<Data>();
            this.ownedArtifacts = new List<Artifact>();
        }
        public static readonly Type DEFAULT_ARTIFACT_TYPE;
        static Artifact() {
            DEFAULT_ARTIFACT_TYPE = new Type();
            DEFAULT_ARTIFACT_TYPE.SetIdentifier("19");
            DEFAULT_ARTIFACT_TYPE.SetName("TERM");
        }
        #endregion

        #region Common OSLC methods
        [OslcDescription("The scope of a resource is a URI for the resource's OSLC Service Provider.")]
        [OslcOccurs(Occurs.ExactlyOne)]
        [OslcPropertyDefinition(OslcConstants.OSLC_CORE_NAMESPACE + "serviceProvider")]
        [OslcRange(OslcConstants.TYPE_SERVICE_PROVIDER)]
        [OslcReadOnly]
        [OslcTitle("Service Provider")]
        public Uri GetServiceProvider() {
            return serviceProvider;
        }
        public void SetServiceProvider(Uri value) {
            this.serviceProvider = value;
        }


        [OslcDescription("A unique identifier for a resource. Assigned by the service provider when a resource is created. Not intended for end-user display.")]
        [OslcOccurs(Occurs.ExactlyOne)]
        [OslcPropertyDefinition(OslcConstants.DCTERMS_NAMESPACE + "identifier")]
        [OslcReadOnly]
        [OslcTitle("Identifier")]
        public string GetIdentifier() {
            return identifier;
        }

        public void SetIdentifier(string identifier) {
            this.identifier = identifier;
        }



        [OslcDescription("The resource type URIs.")]
        [OslcName("type")]
        [OslcPropertyDefinition(OslcConstants.RDF_NAMESPACE + "type")]
        [OslcTitle("Types")]
        public Uri[] GetRdfTypes() {
            return base.GetTypes().ToArray();
        }

        public void SetRdfTypes(Uri[] rdfTypes) {
            base.GetTypes().AddAll(rdfTypes);
        }

        #endregion

        #region Getters & Setters
        [OslcDescription("Artifact Type")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "type")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("Artifact Type")]
        [OslcName("type")]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public Type GetArtifactType() {
            return this.type;
        }

        public void SetArtifactType(Type type) {
            this.type = type;
        }


        [OslcDescription("Content")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "content")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("Content")]
        [OslcName("content")]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public Relationship[] GetRelationships() {
            return this.relationships.ToArray<Relationship>();
        }
        public void SetRelationships(Relationship[] relationships) {
            this.relationships.Clear();
            this.relationships.AddAll(relationships);
        }

        public void SetRelationshipsAsList(List<Relationship> relationships) {
            this.relationships.Clear();
            this.relationships = null;
            this.relationships = relationships;
        }
        public List<Relationship> GetRelationshipsAsList() {
            return this.relationships;
        }
        [OslcDescription("Artifact concept")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "concept")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("Artifact concept")]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public Term GetTerm() {
            return term;
        }
        public void SetTerm(Term term) {
            this.term = term;
        }
        [OslcDescription("The metadata available forthis knowledge element.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "metadata")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The metadata")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public MetaData[] GetMetaData() {
            return this.metadata.ToArray<MetaData>();
        }
        public void SetMetaData(MetaData[] metadata) {
            this.metadata.Clear();
            this.metadata.AddAll(metadata);
        }
        [OslcDescription("The data available forthis knowledge element.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The data")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public Data[] GetData() {
            return this.data.ToArray<Data>();
        }
        public void SetData(Data[] data) {
            this.data.Clear();
            this.data.AddAll(data);
        }
        [OslcDescription("The metadata available for this knowledge element in a list format.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The metadata")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public List<MetaData> GetMetaDataAsList() {
            return this.metadata;
        }
        public void SetMetaDataAsList(List<MetaData> metadata) {
            this.metadata = metadata;
        }
        [OslcDescription("The data available for this knowledge element in a list format.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The data")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public List<Data> GetDataAsList() {
            return this.data;
        }
        public void SetDataAsList(List<Data> data) {
            this.data = data;
        }
        [OslcDescription("The content in binary format for this knowledge element.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The data")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public byte[] GetContent() {
            if (this.Content != null) {
                return Convert.FromBase64String(this.Content);
            } else {
                return null;
            }
        }
        public void SetContent(byte[] artifactContent) {
            if (artifactContent != null) {
                this.Content = Convert.ToBase64String(artifactContent);
            }
        }
        [OslcDescription("The content in image binary format for this knowledge element.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The data")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public byte[] GetSnapshot() {
            if (this.SnapShot != null) {
                return Convert.FromBase64String(this.SnapShot);
            } else {
                return null;
            }
        }
        public void SetSnapshot(byte[] artifactshapshot) {
            if (artifactshapshot != null) {
                this.SnapShot = Convert.ToBase64String(artifactshapshot);
            }

        }
        [OslcDescription("Physical or logical path for this knowledge element.")]
        [OslcPropertyDefinition(SRLShapeConstants.OSLC_KM_SRL_VOCAB + "data")]
        [OslcReadOnly]
        [OslcRepresentation(Representation.Inline)]
        [OslcTitle("The data")]
        [OslcValueShape(OslcConstants.PATH_RESOURCE_SHAPES + "/" + OslcConstants.PATH_PUBLISHER)]
        [OslcValueType(OSLC4Net.Core.Model.ValueType.LocalResource)]
        public string GetPhysicalPath() {
            return this.PhysicalPath;
        }
        public void SetPhysicalPath(string PhysicalPath) {
            this.PhysicalPath = PhysicalPath;
        }
        public bool isKE() {
            return this.ke;
        }
        #endregion

        #region Json Management
        public void Serialize(string path) {
            using (System.IO.StreamWriter file = System.IO.File.CreateText(path)) {
                JsonSerializer serializer = new JsonSerializer();
                serializer.Serialize(file, this);
            }
        }
        public void SerializeToJson(string path) {
            using (System.IO.StreamWriter file = System.IO.File.CreateText(path)) {
                JsonSerializer serializer = new JsonSerializer();
                serializer.Serialize(file, this);
            }
        }
        public string SerializeToJson() {
            string json = string.Empty;
            json = JsonConvert.SerializeObject(this, Formatting.Indented);
            return json;
        }
        public Artifact DeserializeFromJson(string filePath) {
            Artifact result;
            using (System.IO.StreamReader file = System.IO.File.OpenText(filePath)) {
                JsonSerializer serializer = new JsonSerializer();
                result = (Artifact)serializer.Deserialize(file, typeof(Artifact));
            }
            return result;
        }
        public Artifact DeserializeFromJsonString(string json) {
            Artifact result;
            result = JsonConvert.DeserializeObject<Artifact>(json);
            return result;
        }
        #endregion


    }
}